client/TracyProfiler.cpp

```diff
@@ -2889,9 +2897,162 @@ void Profiler::HandleParameter( uint64_t payload )
+// with only executable, not readable, permissions. Typical example (line from /proc/self/maps):
+/*
+746b63b000-746b6dc000 --xp 00042000 07:48 35                             /apex/com.android.runtime/lib64/bionic/libc.so
+*/
+// See https://github.com/wolfpld/tracy/issues/125 .
+// To work around this, we parse /proc/self/maps and we use mprotect to set read permissions
+// on any mappings that contain symbols addresses hit by HandleSymbolCodeQuery.
+
+// Holds some information about a single memory mapping.
+struct MappingInfo {
+    // Start of address range. Inclusive.
+    uintptr_t start_address;
+    // End of address range. Exclusive, so the mapping is the half-open interval
+    // [start, end) and its length in bytes is `end - start`. As in /proc/self/maps.
+    uintptr_t end_address;
+    // Read/Write/Executable permissions.
+    bool perm_r, perm_w, perm_x;
+};
+
+// Internal implementation helper for LookUpMapping(address).
+//
+// Parses /proc/self/maps returning a vector<MappingInfo>.
+// /proc/self/maps is assumed to be sorted by ascending address, so the resulting
+// vector is sorted by ascending address too.
+std::vector<MappingInfo> ParseMappings()
+{
+    std::vector<MappingInfo> result;
+    FILE* file = fopen( "/proc/self/maps", "r" );
+    if( !file ) return result;
+    char line[1024];
+    while( fgets( line, sizeof( line ), file ) )
+    {
+        uintptr_t start_addr;
+        uintptr_t end_addr;
+        if( sscanf( line, "%lx-%lx", &start_addr, &end_addr ) != 2 ) continue;
+        char* first_space = strchr( line, ' ' );
+        if( !first_space ) continue;
+        char* perm = first_space + 1;
+        char* second_space = strchr( perm, ' ' );
+        if( !second_space || second_space - perm != 4 ) continue;
+        result.emplace_back();
+        auto& mapping = result.back();
+        mapping.start_address = start_addr;
+        mapping.end_address = end_addr;
+        mapping.perm_r = perm[0] == 'r';
+        mapping.perm_w = perm[1] == 'w';
+        mapping.perm_x = perm[2] == 'x';
+    }
+    fclose( file );
+    return result;
+}
+
+// Internal implementation helper for LookUpMapping(address).
+//
+// Takes as input an `address` and a
+// known vector `mappings`, and returns a pointer to the MappingInfo
+// describing the mapping that this address belongs to, or nullptr if
+// the address isn't in `mappings`.
+inline MappingInfo* LookUpMapping(std::vector<MappingInfo>& mappings, uintptr_t address)
+{
+    // We assume mappings to be sorted by address, as /proc/self/maps seems to be.
+    // Construct a MappingInfo just for the purpose of using std::lower_bound.
+    MappingInfo needle;
+    needle.start_address = address;
+    needle.end_address = address;
+    // Comparison function for std::lower_bound. Returns true if all addresses in `m1`
+    // are lower than all addresses in `m2`.
+    auto Compare = []( const MappingInfo& m1, const MappingInfo& m2 ) {
```

Comment:
This should be `( const MappingInfo& m1, uintptr_t m2 )`. Then you don't need to construct `needle` struct above, and just pass `address` to `std::lower_bound`.

Reply:
Thanks a lot for this tip, I didn't know about this overload of `std::lower_bound`, the code is simpler this way.